docs/backend/widgets.md

```diff
@@ -12,15 +12,688 @@ myst:
+    </field>
+```
+
+```{note}
+To be included in the XML representation of a schema, you must handle widget parameters through a `WidgetExportImportHandler` utility.
+For a generic interface that handles the attributes, see [`z3c.form.browser.interfaces.IHTMLFormElement`](https://github.com/zopefoundation/z3c.form/blob/42f387d105c305757c8c8c3737c39c960f452acb/src/z3c/form/browser/interfaces.py#L147).
 ```
 
-(backend-widgets-autoform-label)=
+### Set the widget for `z3c.form` plain forms
+
+You can set a field's `widgetFactory` after fields have been declared in a form's class's body.
+
+```{code-block} python
+:emphasize-lines: 23
+:linenos:
+
+import zope.schema
+import zope.interface
+
+from z3c.form import form
+from z3c.form.browser.checkbox import CheckBoxFieldWidget
+
+
+class IReportSchema(zope.interface.Interface):
+    """ Define reporter form fields """
+
+    variables = zope.schema.List(
+        title="Variables",
+        description="Choose which variables to include in the output report",
+        required=False,
+        value_type=zope.schema.Choice(vocabulary="output_variables"))
+
+
+class ReportForm(form.Form):
+    """ A form to output a HTML report from chosen parameters """
+
+    fields = z3c.form.field.Fields(IReportSchema)
+
+    fields["variables"].widgetFactory = CheckBoxFieldWidget
+```
+
+
+### Set a widget dynamically with `Form.updateWidgets()`
+
+You can dynamically set the widget type based on external conditions.
+
+```{code-block} python
+:emphasize-lines: 9
+:linenos:
+
+class EditForm(form.Form):
+
+    label = "Rendering widgets as blocks instead of cells"
+
+    def updateWidgets(self, prefix=""):
+        super().updateWidgets(prefix=prefix)
+
+        # Set a custom widget for a field for this form instance only
+        self.fields["address"].widgetFactory = BlockDataGridFieldFactory
+```
+
+
+## Access and modify widgets
+
+From time to time, you might need to access widgets to get their information or modify them.
+For example, when a user fills out one field, a second field may need to display options dependent on the first field.
+
+
+### Access a widget
+
+You can access a widget by its field's name.
+
+```{code-block} python
+:emphasize-lines: 8
+:linenos:
+
+from z3c.form import form
+
+
+class MyForm(form.Form):
+
+    def update(self):
+        super().update()
+        widget = form.widgets["my_field_name"] # Get one widget
+
+        for w in widget.items(): print(w) # Dump all widgets
+```
+
+To access a widget that's part of a group, you can't use the `updateWidgets()` method.
+The groups and their widgets get initialized after the widgets have been updated.
+Before that, the `groups` attribute is just a list of group factories.
+During the update method though, the groups have been initialized and have their own widget list each.
+To access widgets in a group, you have to access the group in the update method:
+
+``` {code-block} python
+:linenos:
+
+from z3c.form import form
```

Comment:
Why are these added throughout the page? Line numbers don't add anything to these code examples. They just add clutter, and make `{code-block}` and an extra line break unnecessary. If you refer to a specific line number in the narrative text, or emphasize lines, then it makes sense to use the verbose form.

The above three lines (and elsewhere in this page) may be simplified as shown:

````
```python
````

Reply:
didn't notice your commits. fixed again